### PR TITLE
T13439 psr 16 4

### PR DESCRIPTION
**Pull request by niden**
Hello!

* Type: new feature
* Link to issue: #13439 #13672 

**In raising this pull request, I confirm the following:**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist
- [x] I wrote some tests for this PR
- [x] I updated the CHANGELOG

---

- Added `Phalcon\Cache\Cache` class implementing PSR-16. Introducing:
    - `Phalcon\Cache\Adapter`
        - `Phalcon\Cache\Adapter\Apcu`
        - `Phalcon\Cache\Adapter\Libmemcached`
        - `Phalcon\Cache\Adapter\Memory`
        - `Phalcon\Cache\Adapter\Redis`
        - `Phalcon\Cache\Adapter\Stream`
    - `Phalcon\Cache\AdapterFactory`: Factory to create adapters
    - `Phalcon\Cache\CacheFactory`: Factory to create cache objects

- Changed `Phalcon\Cache` namespace and relevant classes to handle storing data to different stores. Introducing:
    - `Phalcon\Storage\Serializer` offering classes that implement the `\Serializable` interface.
        - `Phalcon\Storage\Serializer\Base64`
        - `Phalcon\Storage\Serializer\Igbinary`
        - `Phalcon\Storage\Serializer\Json`
        - `Phalcon\Storage\Serializer\Msgpack`
        - `Phalcon\Storage\Serializer\None`
        - `Phalcon\Storage\Serializer\Php`
        - `Phalcon\Storage\Serializer\SerializerInterface`
    - `Phalcon\Storage\SerializerFactory`: Factory to create serializers
    - `Phalcon\Storage\Adapter` offering classes that implement the `Phalcon\Storage\Adapter\AdapterInterface` interface.
        - `Phalcon\Storage\Adapter\Apcu`
        - `Phalcon\Storage\Adapter\Libmemcached`
        - `Phalcon\Storage\Adapter\Memory`
        - `Phalcon\Storage\Adapter\Redis`
        - `Phalcon\Storage\Adapter\Stream`
    - `Phalcon\Storage\AdapterFactory`: Factory to create adapters

- Removed
    - Caching from the view (simple/view)
        - `Phalcon\Mvc\View::cache()`
        - `Phalcon\Mvc\View::getCache()`
        - `Phalcon\Mvc\View\Simple::cache()`
        - `Phalcon\Mvc\View\Simple::getCache()`
    - Multiple Cache Adapter
        - `Phalcon\Cache\Multiple`
    - Old cache classes
        - `Phalcon\Cache\Backend`
        - `Phalcon\Cache\BackendInterface`
        - `Phalcon\Cache\Backend\Apcu`
        - `Phalcon\Cache\Backend\Factory`
        - `Phalcon\Cache\Backend\File`
        - `Phalcon\Cache\Backend\Libmemcached`
        - `Phalcon\Cache\Backend\Memory`
        - `Phalcon\Cache\Backend\Mongo`
        - `Phalcon\Cache\Backend\Redis`
        - `Phalcon\Cache\Frontend`
        - `Phalcon\Cache\FrontendInterface`
        - `Phalcon\Cache\Frontend\Base64`
        - `Phalcon\Cache\Frontend\Data`
        - `Phalcon\Cache\Frontend\Factory`
        - `Phalcon\Cache\Frontend\Igbinary`
        - `Phalcon\Cache\Frontend\Json`
        - `Phalcon\Cache\Frontend\Msgpack`
        - `Phalcon\Cache\Frontend\None`
        - `Phalcon\Cache\Frontend\Output`

- Added `Str::dirSeparator()` to ensure a path has a trailing slash (not really related to this issue but needed the functionality)

### CODE REVIEW

**Comment by niden** [comment]
cc @phalcon/collaborators @phalcon/core-team

**Comment by niden** [comment]
I will fix the appveyor and regenerate `ext` after this is merged to ensure that old files are removed and new ones added.

**Comment by ruudboon** [comment]
Well done Niden!

File: appveyor.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@
 # For the full copyright and license information, please view
 # the LICENSE.txt file that was distributed with this source code.
 
-version: 4.0.0-alpha.4+{build}
+version: 4.0.0-alpha.5+{build}
```
Comment (sergeyklay) [inline appveyor.yml]: Could you update Version.zep as well

File: phalcon/Cache/AdapterFactory.zep [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     * AdapterFactory constructor.
+     */
+    public function __construct(<SerializerFactory> factory, array! services = [])
+    {
+        var helpers, name, service;
+        
+        let this->serializerFactory = factory;
+
+        let helpers = [
```
Comment (sergeyklay) [inline phalcon/Cache/AdapterFactory.zep]: Why `helpers`? Did you mean `adapters`?

**Review by sergeyklay — DISMISSED** [review]
Great work

File: phalcon/Cache/AdapterFactory.zep (hunk 2 above)
Comment (niden) [reply phalcon/Cache/AdapterFactory.zep]: Fixed

**Pull request merged**